### PR TITLE
Refactor the computation of stack frame parameters

### PR DESCRIPTION
**Pull request by xavierleroy**
The code emitters `$ARCH/emit.mlp` need to know the following characteristics of the Linear code in order to manage stack frames:

1. The maximal size of the stack frame during the function execution [for stack bound checking]
2. Whether the function contains non-tail calls to OCaml functions [for stack bound checking]
3. Whether the function contains anything resembling a non-tail call and requiring the return address to be saved on stack
4. Whether the function needs to allocate a stack frame on entry and free it on return.

Currently, these parameters are computed all over the native back-end: 1 and 2 by function `preproc_stack_check` in `emitaux.ml`, and 4 by functions in `$ARCH/proc.ml`.  The main pain point is 3: `contains_calls` is computed early during the Selection pass, then propagated through later passes, but needs updating in the `Polling` pass.

This PR proposes to group all these computations in the same place: a shared `stackframe_gen.ml` file that is specialized in a per-target `$ARCH/stackframe.ml` file.  The computations are performed late in the native compilation pipeline, so as to minimize the amount of information that needs propagating and updating through the pipeline.

The PR also makes a not-completely-related simplification.  The current code makes provisions to distinguish two conditions: (a) the return address must be saved on stack, and (b) a stack frame must be set up.  (a) implies (b), but a tail function that spills some local variables on the stack needs (b) but not (a).  However, the current code is not taking advantage of this, always saving the return address if a stack frame is set up.  I don't think it matters performance-wise, so I propose to simplify the code accordingly.

### CODE REVIEW

**Comment by xavierleroy** [comment]
Remark: omitting the copyright headers for the new files, this PR actually reduces the code size of the back-end.  That's a good sign for a refactoring.

**Comment by xavierleroy** [comment]
This PR is now ported to all the supported back-ends.  I think it's ready for review.

**Comment by mshinwell** [comment]
I'll review this.

File: asmcomp/arm64/emit.mlp [hunk 1]
```diff
@@ -274,7 +274,7 @@ let emit_stack_adjustment n =
 
 let output_epilogue env f =
   let n = frame_size env in
-  if env.f.fun_contains_calls then
+  if env.f.fun_frame_required then
```
Comment (kayceesrk) [inline asmcomp/arm64/emit.mlp:277]: Looking at this and other replacements of `fun_contains_calls` with `fun_frame_required` in isolation, it is not immediately clear to me why this change is ~correct~ semantically equivalent. In particular, `fun_frame_required` on Arm64 is the one defined in `asmcomp/stackframegen.ml`, which is 

```
method frame_required f contains_calls =
  contains_calls || 
  f.fun_num_stack_slots.(0) > 0 || f.fun_num_stack_slots.(1) > 0
```
which is weaker than `fun_contains_calls`? What am I missing?

Comment (xavierleroy) [reply asmcomp/arm64/emit.mlp:277]: In what sense is it weaker?  Notice that `frame_required` will always return `true` if `contains_calls` is true.

Comment (xavierleroy) [reply asmcomp/arm64/emit.mlp:277]: Oh, by "weaker" you mean "frame_required is true even for some functions that do not contain calls".  This is true, but as I wrote above:
_However, the current code is not taking advantage of [functions that need a frame but don't contain calls], always saving the return address if a stack frame is set up. I don't think it matters performance-wise, so I propose to simplify the code accordingly._

Comment (xavierleroy) [reply asmcomp/arm64/emit.mlp:277]: My reasoning is that a leaf function (containing no calls) that cannot run entirely in registers and requires a stack frame must be pretty big already, so the cost of saving & restoring the return address is negligible.

Comment (kayceesrk) [reply asmcomp/arm64/emit.mlp:277]: Thanks for the clarification. The comment does indeed make sense now that you point it out again :-)

**Review by kayceesrk — APPROVED** [review]
LGTM. This is a much-needed simplification of code.

**Comment by xavierleroy** [comment]
@mshinwell : anything to add to @kayceesrk 's review?

**Comment by mshinwell** [comment]
Sorry I hadn't realised that @kayceesrk was already reviewing this.  This looks good - I still need to check some details, but one minor suggestion so far is to make `Stackframe.analyze` return a record type.

**Comment by mshinwell** [comment]
I've looked through this carefully and it seems correct to me.  It is a nice simplification.

One suggestion: it would be clearer if `contains_nontail_calls` was renamed to `contains_nontail_ocaml_calls` or similar.

Looks like it needs `make alldepend` according to the hygiene check.

**Review by mshinwell — APPROVED** [review]

**Comment by xavierleroy** [comment]
Thanks for the reviews!  After some hemming and humming on where to declare the record type, I went with a record for the result of `Stackframe.analyze`.  Time to merge!
